### PR TITLE
fix: resolve IME enter to complete composition bug

### PR DESCRIPTION
**Pull request by alexisareyn**
Fix race condition where compositionend fires before onKeyDown handler, causing premature search submission when users press Enter to complete character composition.

- Add useIMEComposition hook to track composition state with requestAnimationFrame
- Modify autosuggest to use isComposing() check
- Add comprehensive tests for IME composition scenarios
- Ensure Enter key during composition completion doesn't trigger search

### Description

This PR fixes a critical accessibility issue affecting Korean (and other IME) users where pressing Enter to complete character composition would accidentally trigger search submission instead of just completing the character.

__Problem__: When Korean users type characters like 가 (ga), pressing Enter to complete the character would incorrectly trigger search submission because the browser's `compositionend` event fires before the `onKeyDown` handler executes, creating a race condition.

__Root Cause__: The timing issue meant there was no reliable way to distinguish between "Enter pressed to complete character composition" vs "Enter pressed to trigger search".

Solution: Created a useIMEComposition hook that tracks composition state and uses requestAnimationFrame to maintain the composition flag briefly after compositionend, allowing the onKeyDown handler to correctly identify composition-related Enter keypresses and prevent inappropriate search triggering.

### How has this been tested?

__Automated Testing:__

- Added 13 comprehensive unit tests for the new `useIMEComposition` hook covering character formation, multiple composition cycles, edge cases, and proper cleanup
- Modified existing autosuggest integration tests to verify IME scenarios
- All tests pass: `npm test`

__Manual Testing:__

- Start the dev server: `npm start`

- Navigate to the autosuggest search page at [](http://localhost:8080)<http://localhost:8080>

- Switch to Korean 2-set keyboard input method

- Type Korean characters (e.g., ㄱ + ㅏ to form 가) and press Enter to complete - should NOT trigger search

- Type regular text and press Enter - should trigger search normally

- Run automated tests: `npm test`

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- [N/A] _Changes include appropriate documentation updates._
- [X] _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- [X] _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- [X] _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- [N/A] _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- [X] _Changes are covered with new/existing unit tests?_
- [x] _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Harsh-Anand-Singh** [comment]
Thanks for the contribution! Great PR description the problem & solution breakdown made this easy to review. LGTM 👍

**Review by Harsh-Anand-Singh — APPROVED** [review]

File: src/internal/components/autosuggest-input/index.tsx [hunk 1]
```diff
@@ -180,6 +183,11 @@ const AutosuggestInput = React.forwardRef(
           break;
         }
         case KeyCode.enter: {
+          if (isComposing()) {
```
Comment (georgylobko) [inline src/internal/components/autosuggest-input/index.tsx:186]: Did you try to check `event.detail.shiftKey` and `event.detail.isComposing` instead of introducing this hook?

Comment (alexisareyn) [reply src/internal/components/autosuggest-input/index.tsx:186]: I initially tried `event.detail.isComposing` but it doesn't solve the race condition. The core issue is that `compositionend` fires immediately when Enter is pressed, clearing the composition state *before* `onKeyDown` executes. So by the time we check `event.detail.isComposing`, it's already `false` even for composition-ending Enter keys.

The `useIMEComposition` hook bridges this timing gap by using `requestAnimationFrame` to delay clearing the composition flag, ensuring `onKeyDown` can reliably detect composition-related Enter presses.

File: pages/autosuggest/search.page.tsx [hunk 2]
```diff
@@ -23,6 +23,12 @@ const options: AutosuggestProps.Option[] = [
   { value: '_apple_', label: 'Apple' },
   { value: '_sweet_apple_', label: 'Apple (sweet)', tags: ['sweet'] },
   { value: '_pineapple_', label: 'Pineapple XL', description: 'pine+apple' },
+  // Korean IME test options -- to test, enable Korean 2-set keyboard
+  { value: '_ga_word_', label: '가족 (Family)', description: 'Starts with 가' },
+  { value: '_gak_word_', label: '각도 (Angle)', description: 'Starts with 각' },
+  { value: '_gan_word_', label: '간단 (Simple)', description: 'Starts with 간' },
+  { value: '_gar_word_', label: '가방 (Bag)', description: 'Korean word starting with 가' },
+  { value: '_gae_word_', label: '개념 (Concept)', description: 'Starts with 개' },
```
Comment (georgylobko) [inline pages/autosuggest/search.page.tsx:31]: [non-blocking ] Could also be useful to introduce an integ test in the `__integ__` folder to ensure a real browser behavior

File: src/autosuggest/__tests__/autosuggest.test.tsx [hunk 3]
```diff
@@ -348,6 +348,136 @@ describe('keyboard interactions', () => {
     expect(wrapper.findDropdown()!.findOpenDropdown()).not.toBe(null);
   });
 
+  test('does not select option on enter keydown if part of IME composition', () => {
+    const onChange = jest.fn();
+    const onSelect = jest.fn();
+    const { wrapper, container } = renderAutosuggest(
+      <Autosuggest {...defaultProps} onChange={onChange} onSelect={onSelect} />
+    );
+    const input = container.querySelector('input')!;
```
Comment (georgylobko) [inline src/autosuggest/__tests__/autosuggest.test.tsx]: You can use `wrapper.findNativeInput().getElement()` instead

File: src/autosuggest/__tests__/autosuggest.test.tsx [hunk 4]
```diff
@@ -348,6 +348,136 @@ describe('keyboard interactions', () => {
...
+    wrapper.findNativeInput().keydown(KeyCode.down);
+    input.dispatchEvent(new CompositionEvent('compositionstart'));
+    wrapper.findNativeInput().keydown({ keyCode: KeyCode.enter, isComposing: false });
+
+    expect(onChange).not.toHaveBeenCalled();
+    expect(onSelect).not.toHaveBeenCalled();
+    expect(wrapper.findDropdown().findOpenDropdown()).not.toBe(null);
+  });
+
+  test('selects option on enter keydown when not composing', () => {
```
Comment (georgylobko) [inline src/autosuggest/__tests__/autosuggest.test.tsx]: This scenario is already covered here: https://github.com/cloudscape-design/components/blob/main/src/autosuggest/__tests__/autosuggest.test.tsx#L325

File: src/autosuggest/__tests__/autosuggest.test.tsx [hunk 5]
```diff
@@ -348,6 +348,136 @@ describe('keyboard interactions', () => {
...
+      callback(performance.now());
+      return 1;
+    });
+
+    const onChange = jest.fn();
+    const onSelect = jest.fn();
+    const { wrapper, container } = renderAutosuggest(
+      <Autosuggest {...defaultProps} onChange={onChange} onSelect={onSelect} />
+    );
+    const input = container.querySelector('input')!;
```
Comment (georgylobko) [inline src/autosuggest/__tests__/autosuggest.test.tsx]: Same as above

File: src/autosuggest/__tests__/autosuggest.test.tsx [hunk 6]
```diff
@@ -348,6 +348,136 @@ describe('keyboard interactions', () => {
...
+    expect(onChange).not.toHaveBeenCalled();
+    expect(onSelect).not.toHaveBeenCalled();
+
+    input.dispatchEvent(new CompositionEvent('compositionstart'));
+    input.dispatchEvent(new CompositionEvent('compositionend', { data: '가' }));
+    wrapper.findNativeInput().keydown({ keyCode: KeyCode.enter, isComposing: false });
+    expect(onChange).not.toHaveBeenCalled();
+    expect(onSelect).not.toHaveBeenCalled();
+
+    jest.restoreAllMocks();
```
Comment (georgylobko) [inline src/autosuggest/__tests__/autosuggest.test.tsx]: Could be moved to `beforeEach`

File: src/autosuggest/__tests__/autosuggest.test.tsx [hunk 7]
```diff
@@ -348,6 +348,136 @@ describe('keyboard interactions', () => {
...
+    jest.spyOn(window, 'requestAnimationFrame').mockImplementation(() => {
+      return 1;
+    });
+
+    const onChange = jest.fn();
+    const onSelect = jest.fn();
+    const { wrapper, container } = renderAutosuggest(
+      <Autosuggest {...defaultProps} onChange={onChange} onSelect={onSelect} />
+    );
+    const input = container.querySelector('input')!;
```
Comment (georgylobko) [inline src/autosuggest/__tests__/autosuggest.test.tsx]: Same as above

File: src/internal/hooks/use-ime-composition/__tests__/use-ime-composition.test.tsx [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+import { useIMEComposition } from '../index';
+
+interface Ref {
+  getIsComposing: () => boolean;
+  triggerCompositionStart: () => void;
+  triggerCompositionEnd: (data?: string) => void;
+  getInputElement: () => HTMLInputElement | null;
+}
+
+const Component = React.forwardRef((props, ref: React.Ref<Ref>) => {
```
Comment (georgylobko) [inline src/internal/hooks/use-ime-composition/__tests__/use-ime-composition.test.tsx]: You can use our internal `renderHook` to achieve pretty much the same https://github.com/cloudscape-design/components/blob/main/src/__tests__/render-hook.tsx#L18

File: src/internal/hooks/use-ime-composition/index.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    element.addEventListener('compositionstart', handleCompositionStart);
+    element.addEventListener('compositionend', handleCompositionEnd);
+
+    return () => {
+      element.removeEventListener('compositionstart', handleCompositionStart);
+      element.removeEventListener('compositionend', handleCompositionEnd);
+    };
+  }, [elementRef, handleCompositionStart, handleCompositionEnd]);
+
+  const isComposing = useCallback(() => wasComposingRef.current, []);
```
Comment (georgylobko) [inline src/internal/hooks/use-ime-composition/index.ts]: Here you don't get benefits of using `useCallback` as the ref is a mutable object

File: src/internal/hooks/use-ime-composition/index.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+ * This hook tracks composition lifecycle to prevent the second Enter during the brief window.
+ */
+export function useIMEComposition(elementRef: React.RefObject<HTMLInputElement>) {
+  const wasComposingRef = useRef(false);
+
+  const handleCompositionStart = useCallback(() => {
+    wasComposingRef.current = true;
+  }, []);
+
+  const handleCompositionEnd = useCallback(() => {
```
Comment (georgylobko) [inline src/internal/hooks/use-ime-composition/index.ts]: Here you don't get benefits of using `useCallback` as the ref is a mutable object

File: src/autosuggest/__tests__/autosuggest.test.tsx (hunk 4 above)
Comment (alexisareyn) [reply src/autosuggest/__tests__/autosuggest.test.tsx]: thanks will do

File: pages/autosuggest/search.page.tsx (hunk 2 above)
Comment (alexisareyn) [reply pages/autosuggest/search.page.tsx:31]: will add one with other fixes for comments below, thanks

Comment (alexisareyn) [reply pages/autosuggest/search.page.tsx:31]: had discussion with @georgylobko, there is not a clean way to enable a two set keyboard which would be required to create an accurate integration test, so skipping
